Consul Docs : Add external authorization (ext_authz) support for API gateways (#2776)
hashicorp/web-unified-docs: content/consul/v2.0.x/content/docs/envoy-extension/ext.mdx

          "Prefix": "x-"
        }],
        "DisallowedHeaders": [{
          "Exact": "authorization",
          "IgnoreCase": true
        }],
        "IncludePeerCertificate": true,
        "StatPrefix": "api_gateway_ext_authz"
      }
    }
  }]
}
```

</CodeBlockConfig>
<CodeBlockConfig filename="gateway-service-defaults.yaml">

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceDefaults
metadata:
  name: api-gateway
spec:
  protocol: http
  envoyExtensions:
  - name: builtin/ext-authz
    arguments:
      ProxyType: api-gateway
      Config:
        BootstrapMetadataLabelsKey: envoy.filters.http.header_to_metadata
        GrpcService:
          Target:
            Service:
              Name: authz
              Namespace: default
              Partition: default
            Timeout: 1s
          Authority: authz.example.com
          InitialMetadata:
          - Key: x-consul-gateway
            Value: api-gateway
        AllowedHeaders:
        - Prefix: x-
        DisallowedHeaders:
        - Exact: authorization
          IgnoreCase: true
        IncludePeerCertificate: true
        StatPrefix: api_gateway_ext_authz
```

</CodeBlockConfig>
</CodeTabs>

> After you apply the extension, use the [API gateway configuration entry](/consul/docs/reference/config-entry/api-gateway#extauthz) to set the gateway-wide authorization posture and the [HTTP route `ExtAuthz` filter](/consul/docs/reference/config-entry/http-route#rules-filters-extauthz) to override it for specific routes. For complete workflows, refer to [Delegate API gateway authorization to an external service on virtual machines](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/vm) and [on Kubernetes](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/k8s).

Reviewer: You mention the word `posture` a lot in this PR. 

Before this PR there is no occurrence of the word posture in the documentation. 

* What are we referring to? 
* Is there an external link to point at to get a definition for the word/concept?

It seems a term that is used in identity based system to verify permissions/authorization of services/nodes on the fly, but I want to make sure it is only referring to the general, universally understandable, concept and not to something we are introducing in Consul with different nuances.

Author: "posture" here just means the default authorization behavior (whether ext-authz is enabled or disabled by default), not a new Consul-specific concept. To avoid confusion I'll replace it with plainer wording like "default authorization behavior"/"gateway-wide default" throughout.

> After you apply the extension, use the [API gateway configuration entry](/consul/docs/reference/config-entry/api-gateway#extauthz) to set the gateway-wide authorization posture and the [HTTP route `ExtAuthz` filter](/consul/docs/reference/config-entry/http-route#rules-filters-extauthz) to override it for specific routes. For complete workflows, refer to [Delegate API gateway authorization to an external service on virtual machines](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/vm) and [on Kubernetes](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/k8s).

Reviewer: ```suggestion
After you apply the extension, use the [API gateway configuration entry](/consul/docs/reference/config-entry/api-gateway#extauthz) to set the gateway-wide authorization posture and the [HTTP route `ExtAuthz` filter](/consul/docs/reference/config-entry/http-route#rules-filters-extauthz) to override it for specific routes. Refer to [Delegate API gateway authorization to an external service on virtual machines](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/vm) and [on Kubernetes](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/k8s) for the complete workflows.
```

> After you apply the extension, use the [API gateway configuration entry](/consul/docs/reference/config-entry/api-gateway#extauthz) to set the gateway-wide authorization behavior and the [HTTP route `ExtAuthz` filter](/consul/docs/reference/config-entry/http-route#rules-filters-extauthz) to override it for specific routes. Refer to [Delegate API gateway authorization to an external service on virtual machines](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/vm) and [on Kubernetes](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/k8s) for the complete workflows.

Reviewer: This is not necessary, to apply the system wide behavior only service defaults are needed.

We can clarify this a bit:

```suggestion
The service defaults extension sets the authorization behavior for the whole cluster. If you want to enable external authorization only for specific routes, use the [API gateway configuration entry](/consul/docs/reference/config-entry/api-gateway#extauthz) to disable the global external authorization for the gateway and the [HTTP route `ExtAuthz` filter](/consul/docs/reference/config-entry/http-route#rules-filters-extauthz) to override it for specific routes. 

Refer to [Delegate API gateway authorization to an external service on virtual machines](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/vm) and [on Kubernetes](/consul/docs/north-south/api-gateway/secure-traffic/ext-authz/k8s) for the complete workflows.
```

<Warning>

>   Adding Envoy extensions default proxy configurations may have unintended consequences. We recommend configuring `EnvoyExtensions` in service defaults configuration entries in most cases.

Reviewer: ```suggestion
  Adding Envoy extensions in proxy defaults configuration entries may have unintended consequences. We recommend configuring `EnvoyExtensions` in service defaults configuration entries in most cases.
```
